### PR TITLE
crimson/os/seastore/omap: remove unnecessary extent loads when clearing

### PR DESCRIPTION
**Pull request by aisakaki**
Signed-off-by: Xinyu Huang <xinyu.huang@intel.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc [hunk 1]
```diff
@@ -269,16 +269,28 @@ OMapInnerNode::clear(omap_context_t oc)
...
+	return extent->clear(oc);
+      }).si_then([oc, laddr] {
+	return dec_ref(oc, laddr);
+      }).si_then([ref = OMapNodeRef(this)] {
+	return clear_iertr::now();
+      });
+    } else {
+      return seastar::do_with(
+	std::move(laddr),
+	[oc](auto &laddr) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc:288]: laddr doesn't need `do_with`.

File: src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc [hunk 2]
```diff
@@ -269,16 +269,28 @@ OMapInnerNode::clear(omap_context_t oc)
 {
   LOG_PREFIX(OMapInnerNode::clear);
   DEBUGT("this: {}", oc.t, *this);
-  return trans_intr::do_for_each(iter_begin(), iter_end(), [this, oc] (auto iter) {
+  return trans_intr::do_for_each(iter_begin(), iter_end(),
+    [oc, this] (auto iter) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc:273]: no space between ] and (

File: src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc [hunk 3]
```diff
@@ -269,16 +269,28 @@ OMapInnerNode::clear(omap_context_t oc)
...
+    if (ndepth > 1) {
+      return omap_load_extent(oc, laddr, ndepth).si_then(
+	[oc] (auto &&extent) {
+	return extent->clear(oc);
+      }).si_then([oc, laddr] {
+	return dec_ref(oc, laddr);
+      }).si_then([ref = OMapNodeRef(this)] {
+	return clear_iertr::now();
+      });
+    } else {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc:285]: assert(ndepth == 1)

File: src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc [hunk 4]
```diff
@@ -269,16 +269,28 @@ OMapInnerNode::clear(omap_context_t oc)
...
-      return extent->clear(oc);
-    }).si_then([oc, laddr] {
-      return dec_ref(oc, laddr);
-    }).si_then([ref = OMapNodeRef(this)] {
-      return clear_iertr::now();
-    });
+    auto ndepth = get_meta().depth - 1;
+    if (ndepth > 1) {
+      return omap_load_extent(oc, laddr, ndepth).si_then(
+	[oc] (auto &&extent) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc:278]:
```
return omap_load_extent(oc, laddr, ndepth
).si_then([oc](auto &&extent) {
```

Comment (aisakaki) [reply src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc]: all addressed.

**Review by cyx1231st — APPROVED** [review]

**Comment by aisakaki** [comment]
jenkins test make check

**Review by liu-chunmei — APPROVED** [review]

**Pull request merged**